src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -107,36 +123,119 @@ class BillingTxCountMetricConsumerStrategy(ProcessingStrategy[KafkaPayload]):
-        self.__last_commit = datetime.now()
-        self.__ready_to_commit: MutableMapping[Partition, Position] = {}
-        self.__closed = False
+        self._closed: bool = False
+        self._commit = commit
+
+        self._billing_topic = Topic(settings.KAFKA_OUTCOMES_BILLING)
+        self._billing_producer = self._get_billing_producer()
+        # Especially on incidents generating big backlogs, we must expect a lot
+        # of removals from the beginning of a big FIFO. A dequeue provides O(1)
+        # time on removing items from the beginning; while a regular list takes
+        # O(n).
+        self._ongoing_billing_outcomes: Deque[MetricsBucketBilling] = deque()
+
+    def _get_billing_producer(self) -> KafkaProducer:
+        servers = _get_bootstrap_servers(topic=self._billing_topic.name, force_cluster=None)
+        return KafkaProducer(
+            build_kafka_configuration(default_config={}, bootstrap_servers=servers)
+        )
 
     def poll(self) -> None:
-        if self._should_commit():
-            self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            self._process_metrics_billing_bucket(timeout=None)
+
+    def _process_metrics_billing_bucket(self, timeout: Optional[float] = None) -> None:
+        """
+        Takes the first billing outcome from the queue and if it's completed,
+        commits the metrics bucket.
+
+        When a future has thrown an exception, logs it and commits the bucket
+        anyway. wait_time is used to wait on the future to finish; if `None` is
+        provided and the future is not completed, no action is taken.
+        """
+        if len(self._ongoing_billing_outcomes) < 1:
+            return
+
+        billing_future, metrics_msg = self._ongoing_billing_outcomes[0]
+        if billing_future:
+            try:
+                if not timeout and not billing_future.done():
+                    return
+                billing_future.result(timeout)
+            except Exception:
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=billing_future.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
+
+        self._ongoing_billing_outcomes.popleft()
+        self._commit({metrics_msg.partition: Position(metrics_msg.next_offset, datetime.now())})
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+
+        while self._ongoing_billing_outcomes:
+            now = time.time()
+            time_left = deadline - now if deadline else None
+
+            if time_left is not None and time_left <= 0:
+                logger.warning(
+                    f"join timed out, items left in the queue: {len(self._ongoing_billing_outcomes)}"
+                )
+                break
+
+            self._process_metrics_billing_bucket(time_left)
+
```

Comment:
Please call `self._commit({}, force=True)` here. Unfortunately it can't happen automatically. This tells arroyo to ignore the policy rules and just commit all the latest staged offsets.